### PR TITLE
Geojson features

### PR DESCRIPTION
**Pull request by timoore**
## Description
Add feature IDs to primitives derived from GeoJSON features. These don't refer to metadata yet.

Also, add a test for holes in GeoJSON polygons.

## Issue number or link
Addresses #1353 and #1354.

## Author checklist

- [ x] I have submitted a [Contributor License Agreement](https://github.com/CesiumGS/community/tree/main/CLAs) (only needed once).
- [ x] I have done a full self-review of my code.
- [ ] I have updated `CHANGES.md` with a short summary of my change (for user-facing changes).
- [ ] I have added or updated unit tests to ensure consistent code coverage as necessary.
- [ ] I have updated the documentation as necessary.

## Reviewer checklist

Thank you for taking the time to review this PR. By approving a PR you are taking as much responsibility for these changes as the author.

As you review, please go through the checklist below:

- [x] Review and run all parts of the test plan on this branch and verify it matches expectations.
    - [ ] If the issue is a bug please make sure you can reproduce the bug in the main branch and then checkout this branch to make sure it actually solved the issue.
- [x] Review the code and make sure you do not have any remaining questions or concerns. You should understand the code change and the chosen approach. If you are not confident or have doubts about the code, please do not hesitate to ask questions.
    - [x] Code should follow the [C++ Style Guide](https://cesium.com/learn/cesium-native/ref-doc/style-guide.html)
- [x] Review the unit tests and make sure there are no missing tests or edge cases.
- [x] Review documentation changes and updates to `CHANGES.md` to make sure they accurately cover the work in this PR.
- [x] Verify that the [Contributor License Agreement](https://github.com/CesiumGS/community/tree/main/CLAs) has been submitted, if needed.

### CODE REVIEW

File: Cesium3DTilesSelection/test/TestTilesetHeightQueries.cpp [hunk 1]
```diff
@@ -356,4 +356,28 @@ TEST_CASE("Tileset height queries") {
...
+
+    while (!future.isReady()) {
+      externals.asyncSystem.dispatchMainThreadTasks();
+      tileset.loadTiles();
+    }
+
+    SampleHeightResult results = future.waitInMainThread();
+    CHECK(results.sampleSuccess[0]);
+    CHECK(!results.sampleSuccess[1]);
+  }
```
Comment (j9liu) [inline Cesium3DTilesSelection/test/TestTilesetHeightQueries.cpp:382]: Is this for https://github.com/CesiumGS/cesium-native/issues/1353? Seems unrelated to this PR.

File: CesiumVectorData/src/GltfConverter.cpp [hunk 2]
```diff
@@ -42,36 +44,215 @@ using namespace CesiumVectorData;
...
+  finalizePrimitive(const std::vector<uint32_t>& elements, int32_t mode);
+  void preparePositions();
   // Create a buffer view from a whole buffer
   int32_t makeBufferView(int32_t buffer, int32_t target);
+  // A more nuanced view
+  int32_t makeBufferView(
+      int32_t buffer,
+      int32_t target,
+      int64_t byteOffset,
+      int64_t byteLength);
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp:146]: Why not just use this `makeBufferView` function and delete the other? Better to be over-specific anyways.

File: CesiumVectorData/src/GltfConverter.cpp [hunk 3]
```diff
@@ -42,36 +44,215 @@ using namespace CesiumVectorData;
...
+      const GeoJsonMultiPoint& object,
+      const GeoJsonObject* pFeature) {
+    for (const auto& cartoDegrees : object.coordinates) {
+      doPoint(cartoDegrees, pFeature);
+    }
+  }
+
+  std::unordered_map<const GeoJsonObject*, uint32_t> gltfFeatureIds;
+  std::vector<const GeoJsonObject*> geoJsonFeatures;
+
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: Move this to the top with the other fields? Confusing because this gets referenced in the constructor before it even appears in the struct definition.

File: CesiumVectorData/src/GltfConverter.cpp
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp:1]: Please update CHANGES.md since this is a user-facing change.

File: CesiumVectorData/src/GltfConverter.cpp [hunk 4]
```diff
@@ -42,36 +44,215 @@ using namespace CesiumVectorData;
...
       int64_t byteOffset,
       int64_t count,
       int32_t componentType,
       const std::string& accessorType);
+  void expandRegion(const glm::dvec3& cartoDegrees) {
+    this->regionBuilder.expandToIncludePosition(Cartographic::fromDegrees(
+        cartoDegrees.x,
+        cartoDegrees.y,
+        cartoDegrees.z));
+  };
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: **NIT**: This function isn't doing much; it's only used once in `addCoordinateDegrees` and is just a wrapper around `regionBuilder`. I would remove.

File: CesiumVectorData/src/GltfConverter.cpp [hunk 5]
```diff
@@ -42,36 +44,215 @@ using namespace CesiumVectorData;
...
+  template <typename GeoJsonType>
+  void visitGeometry(const GeoJsonType&, const GeoJsonObject*) {}
+
+  void doLinestring(
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: Opt for more descriptive function names:

```suggestion
  void createLinestring(
```

File: CesiumVectorData/src/GltfConverter.cpp [hunk 6]
```diff
@@ -42,36 +44,215 @@ using namespace CesiumVectorData;
...
+      const std::vector<glm::dvec3>& lineStringCoords,
+      const GeoJsonObject* pFeature) {
+    auto stringIndices = lineStringToLines(
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: Per [style guide](https://cesium.com/learn/cesium-native/ref-doc/style-guide.html#auto),

> Avoid using auto except in the following cases where writing the full type is burdensome or impossible

```suggestion
    std::vector<uint32_t> stringIndices = lineStringToLines(
```

File: CesiumVectorData/src/GltfConverter.cpp [hunk 7]
```diff
@@ -42,36 +44,215 @@ using namespace CesiumVectorData;
...
+      const GeoJsonObject* pFeature) {
+    for (const auto& lineStringCoords : object.coordinates) {
+      doLinestring(lineStringCoords, pFeature);
+    }
+  }
+
+  using PolygonRing = std::vector<glm::dvec3>;
+  using Polygon = std::vector<PolygonRing>;
+
+  void doPolygon(const Polygon& polygonRings, const GeoJsonObject* pFeature) {
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: Opt for more descriptive function name.

```suggestion
  void createPolygon(const Polygon& polygonRings, const GeoJsonObject* pFeature) {
```

File: CesiumVectorData/src/GltfConverter.cpp [hunk 8]
```diff
@@ -42,36 +44,215 @@ using namespace CesiumVectorData;
...
+    uint32_t elementBase = uint32_t(this->globalCoordinates.size());
+    uint32_t featureId = getFeatureId(pFeature);
+    for (const auto& contour : polygonRings) {
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: We already have a type alias for this, why not use that?

```suggestion
    for (const PolygonRing& contour : polygonRings) {
```

File: CesiumVectorData/src/GltfConverter.cpp [hunk 9]
```diff
@@ -42,36 +44,215 @@ using namespace CesiumVectorData;
...
+
+  void visitGeometry(
+      const GeoJsonMultiPolygon& object,
+      const GeoJsonObject* pFeature) {
+    for (const std::vector<PolygonRing>& polygonRings : object.coordinates) {
+      doPolygon(polygonRings, pFeature);
+    }
+  }
+
+  void doPoint(const glm::dvec3& cartoDegrees, const GeoJsonObject* pFeature) {
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: Opt for more descriptive function naming:

```suggestion
  void createPoint(const glm::dvec3& cartoDegrees, const GeoJsonObject* pFeature) {
```

File: CesiumVectorData/src/GltfConverter.cpp [hunk 10]
```diff
@@ -42,36 +44,215 @@ using namespace CesiumVectorData;
...
+
+  void
+  visitGeometry(const GeoJsonPoint& object, const GeoJsonObject* pFeature) {
+    doPoint(object.coordinates, pFeature);
+  }
+
+  void visitGeometry(
+      const GeoJsonMultiPoint& object,
+      const GeoJsonObject* pFeature) {
+    for (const auto& cartoDegrees : object.coordinates) {
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp:231]: Per [style guide](https://cesium.com/learn/cesium-native/ref-doc/style-guide.html#auto),

> Avoid using auto except in the following cases where writing the full type is burdensome or impossible

```suggestion
    for (const glm::dvec3& cartoDegrees : object.coordinates) {
```

File: CesiumVectorData/src/GltfConverter.cpp [hunk 11]
```diff
@@ -92,431 +273,169 @@ int32_t GltfConverterImpl::makeAccessor(
...
-  // The double-precision positions must be converted to single-precision to
-  // conform to the glTF spec.
-  glm::vec3* pPosition = reinterpret_cast<glm::vec3*>(positionBytes.data());
-  for (const glm::dvec3& localPosition : localPositions) {
-    *pPosition++ = glm::vec3(localPosition.x, localPosition.y, localPosition.z);
+  size_t numCoords = this->globalCoordinates.size();
+  positionBuffer.cesium.data.resize(sizeof(glm::vec3) * numCoords);
+  glm::vec3* const position32Base =
+      reinterpret_cast<glm::vec3*>(positionBuffer.cesium.data.data());
+  glm::vec3* position32 = position32Base;
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: Prefix pointers with `p` per style guide:

```suggestion
  glm::vec3* const pPosition32Base =
      reinterpret_cast<glm::vec3*>(positionBuffer.cesium.data.data());
  glm::vec3* pPosition32 = pPosition32Base;
```

File: CesiumVectorData/src/GltfConverter.cpp [hunk 12]
```diff
@@ -92,431 +273,169 @@ int32_t GltfConverterImpl::makeAccessor(
...
+  for (const auto& position : this->globalCoordinates) {
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp:391]: Per [style guide](https://cesium.com/learn/cesium-native/ref-doc/style-guide.html#auto),

> Avoid using auto except in the following cases where writing the full type is burdensome or impossible

```suggestion
  for (const glm::dvec3& position : this->globalCoordinates) {
```

File: CesiumVectorData/src/GltfConverter.cpp [hunk 13]
```diff
@@ -92,431 +273,169 @@ int32_t GltfConverterImpl::makeAccessor(
...
+      sizeof(uint32_t) * numCoords);
+  this->featureIdBufferViewIndex = makeBufferView(
+      this->featureIdBufferIndex,
+      BufferView::Target::ARRAY_BUFFER);
+  this->featureIdAccessorIndex = makeAccessor(
+      this->featureIdBufferViewIndex,
       0,
-      int64_t(indices.size()),
+      int64_t(numCoords),
       Accessor::ComponentType::UNSIGNED_INT,
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: Would be nice if the chosen component type scaled to the number of features _actually_ present in the GeoJSON. That way if you have a small number of features, you could use uint8 instead of uint32. But that can be out-of-scope for this PR.

File: CesiumVectorData/src/GltfConverter.cpp [hunk 14]
```diff
@@ -92,431 +273,169 @@ int32_t GltfConverterImpl::makeAccessor(
...
+  MeshPrimitive& primitive = primitiveMesh.primitives.emplace_back();
+  primitive.attributes["POSITION"] = this->positionAccessorIndex;
+  primitive.attributes["_FEATURE_ID_0"] = this->featureIdAccessorIndex;
+  primitive.indices = elementAccessorIndex;
+  primitive.mode = mode;
+  primitive.material = 0;
+  ExtensionExtMeshFeatures& extension =
+      primitive.addExtension<ExtensionExtMeshFeatures>();
+  FeatureId& featureId = extension.featureIds.emplace_back();
+  featureId.attribute = this->featureIdAccessorIndex;
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: `attribute` does not refer to the accessor index, it refers to the _set index_ of the features to construct `_FEATURE_ID_n`.  See the spec:

https://github.com/CesiumGS/glTF/blob/cesium-native/extensions/2.0/Vendor/EXT_mesh_features/schema/featureIdAttribute.schema.json


```suggestion
  featureId.attribute = 0;
```

File: CesiumVectorData/src/GltfConverter.cpp [hunk 15]
```diff
@@ -92,431 +273,169 @@ int32_t GltfConverterImpl::makeAccessor(
...
+  featureId.featureCount = int64_t(this->gltfFeatureIds.size());
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: `featureCount` does not refer to the number of features in the entire model; it refers to the number of features within the attribute. See the spec:

https://github.com/CesiumGS/glTF/blob/cesium-native/extensions/2.0/Vendor/EXT_mesh_features/schema/featureId.schema.json

File: CesiumVectorData/src/GltfConverter.cpp [hunk 16]
```diff
@@ -42,36 +44,215 @@ using namespace CesiumVectorData;
...
+      ringCopy.push_back(Point{ring[i][0], ring[i][1]});
+    }
+  }
+  return mapbox::earcut<uint32_t>(polygon);
+}
+
 struct GltfConverterImpl {
   const GeoJsonDocument& geoJson;
   Model model;
+  // Accumlated as the data is traversed, then used to transform coordinates
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]:
```suggestion
  // Accumulated as the data is traversed, then used to transform coordinates
```

File: CesiumVectorData/src/GltfConverter.cpp
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp:1]: Also, I think it should be straightforward to write unit tests for `GltfConverter`, especially since we already have sample data in-repo. Please add them before the next review!

**Review by j9liu — CHANGES_REQUESTED** [review]
Thanks for the PR @timoore, I have several comments that need to get addressed. Can you please make sure to test this in Cesium for Unreal before the next review? I caught a bug that prevents the feature IDs from working.

File: Cesium3DTilesSelection/test/TestTilesetHeightQueries.cpp (hunk 1 above)
Comment (timoore) [reply Cesium3DTilesSelection/test/TestTilesetHeightQueries.cpp:382]: Yeah, I mentioned that it was small enough that I wasn't going to file a separate PR for it.

File: CesiumVectorData/src/GltfConverter.cpp (hunk 15 above)
Comment (timoore) [reply CesiumVectorData/src/GltfConverter.cpp]: Strictly speaking that is true, but in our existing code we don't bother to determine the precise number of unique features in an attribute. See e.g. `BatchTableToGltfStructuralMetadata::convertFromB3dm` in BatchTableToGltfStructuralMetadata.cpp.

File: CesiumVectorData/src/GltfConverter.cpp (hunk 14 above)
Comment (timoore) [reply CesiumVectorData/src/GltfConverter.cpp]: Whoops!

File: CesiumVectorData/src/GltfConverter.cpp (hunk 2 above)
Comment (timoore) [reply CesiumVectorData/src/GltfConverter.cpp:146]: I prefer the shorter version.

File: CesiumVectorData/src/GltfConverter.cpp (hunk 13 above)
Comment (timoore) [reply CesiumVectorData/src/GltfConverter.cpp]: I think it's out of scope and, given that this is for used for returning data to client code, probably debatable as an optimization.

File: CesiumVectorData/src/GltfConverter.cpp (hunk 5 above)
Comment (timoore) [reply CesiumVectorData/src/GltfConverter.cpp]: It's not really creating a linestring, it's translating a linestring to lines. Would you prefer `processLinestring`?

File: CesiumVectorData/src/GltfConverter.cpp (hunk 10 above)
Comment (timoore) [reply CesiumVectorData/src/GltfConverter.cpp:231]: I'll change this if you really want, but I feel this is close enough to the iterator case to justify the convenience of using `auto`.

File: CesiumVectorData/src/GltfConverter.cpp (hunk 5 above)
Comment (j9liu) [reply CesiumVectorData/src/GltfConverter.cpp]: Sure, that's better.

File: CesiumVectorData/src/GltfConverter.cpp
Comment (timoore) [reply CesiumVectorData/src/GltfConverter.cpp:1]: I will add tests, but unfortunately all our test data only has one feature per GeoJSON file!

File: CHANGES.md [hunk 17]
```diff
@@ -5,13 +5,15 @@
 ##### Additions :tada:
 
 - Added support for [`EXT_mesh_primitive_edge_visibility`](https://github.com/KhronosGroup/glTF/pull/2479) in `CesiumGltf`, `CesiumGltfReader`, and `CesiumGltfWriter`.
+- Added support for [`3DTILES_content_conditional`](https://github.com/CesiumGS/3d-tiles/pull/834) in `Cesium3DTiles`, `Cesium3DTilesReader`, and `Cesium3DTilesWriter`.
+- Added support for feature ids in glTF translated from `MAXAR_geojson_content` tiles.
```
Comment (j9liu) [inline CHANGES.md]:
```suggestion
- Added support for feature IDs in glTF translated from `MAXAR_geojson_content` tiles.
```

File: tools/generate-classes/3dTiles.json [hunk 18]
```diff
@@ -28,11 +28,26 @@
...
+      "overrideName": "Extension3dTilesContentConditional"
+    },
+    "Content for 3DTILES_content_conditional": {
+      "overrideName": "ConditionalContent"
+    },
+    "One item of the conditional content for 3DTILES_content_conditional": {
+      "overrideName": "ConditionalContentItem"
+    },
+    "MAXAR_content_geojson tileset extension": {
+      "overrideName": "ExtensionMaxarContentGeojson"
```
Comment (j9liu) [inline tools/generate-classes/3dTiles.json]: Not trying to nitpick, but our other classes are titled `GeoJson___` whereas this is `Geojson`. Can we be consistent?

```suggestion
      "overrideName": "ExtensionMaxarContentGeoJson"
```

File: CHANGES.md (hunk 17 above)
Comment (j9liu) [reply CHANGES.md]: Not trying to nitpick, but we do capitalize "ID" everywhere public-facing, e.g., https://github.com/CesiumGS/glTF/tree/cesium-native/extensions/2.0/Vendor/EXT_mesh_features#feature-ids

File: CesiumVectorData/src/GltfConverter.cpp [hunk 19]
```diff
@@ -404,21 +406,44 @@ void GltfConverterImpl::preparePositions() {
...
+    std::memcpy(
+        featureIdBuffer.cesium.data.data(),
+        this->featureIds.data(),
+        sizeof(uint32_t) * numCoords);
+  } else {
+    componentType = Accessor::ComponentType::FLOAT;
+    float* featureDest =
+        reinterpret_cast<float*>(featureIdBuffer.cesium.data.data());
+    for (size_t i = 0; i < numCoords; ++i) {
+      featureDest[i] = static_cast<float>(this->featureIds[i]);
```
Comment (j9liu) [inline CesiumVectorData/src/GltfConverter.cpp]: `p` prefix for pointers:

```suggestion
    float* pFeatureDest =
        reinterpret_cast<float*>(featureIdBuffer.cesium.data.data());
    for (size_t i = 0; i < numCoords; ++i) {
      pFeatureDest[i] = static_cast<float>(this->featureIds[i]);
```

File: CesiumVectorData/src/GltfConverter.cpp (hunk 15 above)
Comment (j9liu) [reply CesiumVectorData/src/GltfConverter.cpp]: The main place where "we don't bother" is because it would be really inefficient to iterate over the batch ID buffer:
https://github.com/CesiumGS/cesium-native/blob/dc3f3ca363576422be58bfe7ebee12909aa7740c/Cesium3DTilesContent/src/BatchTableToGltfStructuralMetadata.cpp#L1928-L1930

Otherwise, I believe we do set this number accurately for points, instances, etc. based on other variables that may be present in the legacy files. It's surprising to not do it here for GeoJSON when we have full control over what we are creating.

**Review by j9liu — COMMENTED** [review]
Thanks for the updates @timoore ! This is looking really close, just a few small comments before merge.

File: tools/generate-classes/3dTiles.json (hunk 18 above)
Comment (timoore) [reply tools/generate-classes/3dTiles.json]: I certainly don't mind changing it, but I was trying to be consistant with our translation of glTF and 3D Tiles extension names to C++, which until now has capitalized the words of the extension name and nothing else... but there is "3DTILES" -> "3dTiles", so maybe there isn't actually a strict rule.

**Comment by timoore** [comment]
@j9liu This is ready for another look. A bug having to do with the feature id buffer view byte stride has been fixed, in addition to your own comments.

File: CHANGES.md (hunk 17 above)
Comment (j9liu) [reply CHANGES.md]: @timoore did you miss this comment?

File: tools/generate-classes/3dTiles.json (hunk 18 above)
Comment (j9liu) [reply tools/generate-classes/3dTiles.json]: @timoore Let's change it, and if we want to revisit these conventions we can do so at a later time

**Comment by j9liu** [comment]
@timoore The changes look good, but I think you missed two of my other comments.

File: CHANGES.md [hunk 20]
```diff
@@ -6,7 +6,7 @@
 
 - Added support for [`EXT_mesh_primitive_edge_visibility`](https://github.com/KhronosGroup/glTF/pull/2479) in `CesiumGltf`, `CesiumGltfReader`, and `CesiumGltfWriter`.
 - Added support for [`3DTILES_content_conditional`](https://github.com/CesiumGS/3d-tiles/pull/834) in `Cesium3DTiles`, `Cesium3DTilesReader`, and `Cesium3DTilesWriter`.
-- Added support for feature ids in glTF translated from `MAXAR_geojson_content` tiles.
+- Added support for feature IDs in glTF translated from `MAXAR_geojson_content` tiles.
```
Comment (j9liu) [inline CHANGES.md]: Sigh, this slipped my last review...

```suggestion
- Added support for feature IDs in glTF translated from `MAXAR_content_geojson` tiles.
```

**Comment by j9liu** [comment]
Thanks @timoore !
